### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by dzhekzzzz**
changed many bugs
added random led timer generator
improved work stability

### CODE REVIEW

File: LEDGAME.ino [hunk 1]
```diff
@@ -3,7 +3,6 @@
 int a = 1;
```
Comment (Antdigapony) [inline LEDGAME.ino]: change this variable name from 'a' to 'gameState' or 'gameStatus'

Comment (Antdigapony) [reply LEDGAME.ino]: also I recommend to use the directives to the pre-processor:

#define GAME_SET_UP (1)
#define GAME_START (2)
#define GAME_RUNNING (3)
#define GAME_END (4)

and set the former 'a' variable gameState:

int gameState = GAME_SET_UP;

File: LEDGAME.ino [hunk 2]
```diff
@@ -35,30 +33,26 @@ class Timer
...
-  {
-     buttWait(); // it is not right to do so - LedTImer mus be a led timer - it do not have to deal with buttons. You better run buttonWait in the loop()
-  }
 };
 
 Timer timer; 
 Timer ledTimer;
 
+
 bool isLedTimerRun()
```
Comment (Antdigapony) [inline LEDGAME.ino]: send this func to the class Timer
or use Timer-class function instead this func

File: LEDGAME.ino [hunk 3]
```diff
@@ -149,47 +125,60 @@ int buttWait(int num)
...
         }
-        timer.updateTimer(); // обновляем таймер
-        timer.isTimerRun();
-        random_led(b);
-        ledTimer.setTimer(500); //ставим сколько дает времени на нажатие
+        timer.updateTimer(); // updating timer valuerun
+        random_led(b); //turn on random led
+        ledTimer.setTimer(t); //setting time for leds
+        timer.i_oldLedTimer = millis();
+        while ( isLedTimerRun() == true )
```
Comment (Antdigapony) [inline LEDGAME.ino]: recommendations: something like that:

while(ledTimer.isTimerRunning())
{
   if (buttonIsPressed()) 
   {
       break;
   }
}

File: LEDGAME.ino [hunk 4]
```diff
@@ -149,47 +125,60 @@ int buttWait(int num)
...
+          if (digitalRead(buttInit[b]) == HIGH) 
+           {
+            butt[b] = b;
+             break;
+           }
+        }
         buttWait(b);
-        clearbutt(); //очищаю значение кнопок  
-        led_off(); //выключаем все лампочки
+        clearbutt(); //clearing butt bufer
```
Comment (Antdigapony) [inline LEDGAME.ino]: it's better to send this func with scores to the GAME_OVER case or to the separate state/case SHOW_SCORES

**Review by Antdigapony — CHANGES_REQUESTED** [review]

File: LEDGAME.ino [hunk 5]
```diff
@@ -1,12 +1,14 @@
...
 #include <stdio.h> 
-int a = 1;
+int gameState = 1;
 int score = 0;
 int noButtPress = 0;
-int num;
 int buttInit[4] = {4, 7, 10, 12} ;
 int ledInit[4] = {3, 9, 5, 2} ;
 int butt[5];
+#define GAME_SET_N_START 1
```
Comment (Antdigapony) [inline LEDGAME.ino]: just a small remark - use () - round brackets to avoid mistakes
#define GAME_SET_IN_START (1)

File: LEDGAME.ino [hunk 6]
```diff
@@ -20,127 +22,95 @@ void setup()
       } 
 }
 
-
 class Timer
 {
   public:
   unsigned long  i_timer;
   unsigned long  i_ledTimer;
```
Comment (Antdigapony) [inline LEDGAME.ino]: this class-field or class-variable is not used anywhere - maybe we should delete it?

File: LEDGAME.ino [hunk 7]
```diff
@@ -20,127 +22,95 @@ void setup()
...
 
-
 class Timer
 {
   public:
   unsigned long  i_timer;
   unsigned long  i_ledTimer;
   unsigned long  i_oldLedTimer;
   unsigned long  limitValue;
-  unsigned long old_i_timer;
```
Comment (Antdigapony) [inline LEDGAME.ino:31]: and this too -

File: LEDGAME.ino [hunk 8]
```diff
@@ -20,127 +22,95 @@ void setup()
       } 
 }
 
-
 class Timer
 {
   public:
   unsigned long  i_timer;
```
Comment (Antdigapony) [inline LEDGAME.ino]: it's better to use ONE STYLE for the variables:
camelCase without _ lowerspaces
and I propose that variables should also "tell" something about themselves:

unsigned long currentTime;
unsigned long previousTime;

File: LEDGAME.ino [hunk 9]
```diff
@@ -20,127 +22,95 @@ void setup()
...
 class Timer
 {
   public:
   unsigned long  i_timer;
   unsigned long  i_ledTimer;
   unsigned long  i_oldLedTimer;
   unsigned long  limitValue;
-  unsigned long old_i_timer;
+  unsigned long  old_i_timer;
+  unsigned long  newTimeLed;
```
Comment (Antdigapony) [inline LEDGAME.ino]: this variable is also un-used

File: LEDGAME.ino [hunk 10]
```diff
@@ -20,127 +22,95 @@ void setup()
...
+//generating random time for led
+int randomGenTime()
 {
-   digitalWrite(ledInit[num], HIGH);   
+  int randomtime = (rand() % 9 + 3) * 100; 
+  return randomtime;
 }
 
+ //turn on random led
+int randomLed(int num)
```
Comment (Antdigapony) [inline LEDGAME.ino]: better call it
randomLedOn - cause we are turning the random led on

File: LEDGAME.ino [hunk 11]
```diff
@@ -20,127 +22,95 @@ void setup()
...
-   digitalWrite(ledInit[3], LOW);
-   digitalWrite(ledInit[0], LOW);
+      for (int l = 0; l < 4; l++)
+      {
+        digitalWrite(ledInit[l], LOW);
+      } 
   }  
 
 //выдаем результат в зависимости от нажатой кнопки 
 int buttWait(int num)
```
Comment (Antdigapony) [inline LEDGAME.ino]: proposal to delete this function
it is redundant

Comment (Antdigapony) [reply LEDGAME.ino]: instead - create a function:

bool checkButtonPressed(int num)
{
    return (digitalRead(buttInit[num]) == HIGH)
}

Comment (Antdigapony) [reply LEDGAME.ino]: you can launch this function in a cycle and use for the ALL buttons check one by one

File: LEDGAME.ino [hunk 12]
```diff
@@ -149,52 +119,74 @@ int buttWait(int num)
       }
     }
 
-//ждем кнопку старт
-void buttsWait()
-{
-  while (butt[0] == 0) 
+ //waiting for start button
+void startButtonWait()
```
Comment (Antdigapony) [inline LEDGAME.ino:137]: now you can use:
while (!checkButtonPressed(0))

File: LEDGAME.ino [hunk 13]
```diff
@@ -149,52 +119,74 @@ int buttWait(int num)
...
+        ledTimer.setTimer(t); //setting time for leds
+        if ( (timeleft) <= 999)
+         {
+          t = timeleft;    
+          gameState = GAME_END;     
+         }
+        while ( ledTimer.isTimerRunning() == true )
         {
-        a = 3;
+          if (digitalRead(buttInit[b]) == HIGH)
```
Comment (Antdigapony) [inline LEDGAME.ino]: you can use checkButtonPressed here too

File: LEDGAME.ino [hunk 14]
```diff
@@ -149,52 +119,74 @@ int buttWait(int num)
...
-        clearbutt(); //очищаю значение кнопок  
-        led_off(); //выключаем все лампочки
+        clearButt(); //clearing butt bufer
+        ledOff(); //turn off all leds
         Serial.print("                                                              your score is ");
         Serial.print(score);
         Serial.print("\n \n");
-        delay(500);
-
+        waitTimer.setTimer(t);
```
Comment (Antdigapony) [inline LEDGAME.ino:191]: waitTimer - were do you set this Timer and were do you update this Timer ?

File: LEDGAME.ino [hunk 15]
```diff
@@ -149,52 +119,74 @@ int buttWait(int num)
...
+           {
+            butt[b] = b;
+             break;
+           }
         }
-        timer.updateTimer(); // обновляем таймер
-        timer.isTimerRun();
-        random_led(b);
-        ledTimer.setTimer(500); //ставим сколько дает времени на нажатие
         buttWait(b);
```
Comment (Antdigapony) [inline LEDGAME.ino]: this is redundant or can be written here - to compare the button and buttonInit

File: LEDGAME.ino (hunk 14 above)
Comment (dzhekzzzz) [reply LEDGAME.ino:191]: setting timer by parameter t, function updating itself in func "waitTimer.isTimerRunning()" below

File: LEDGAME.ino [hunk 16]
```diff
@@ -1,12 +1,15 @@
...
-int a = 1;
+int i = 0;
+int gameState = 1;
 int score = 0;
 int noButtPress = 0;
-int num;
 int buttInit[4] = {4, 7, 10, 12} ;
 int ledInit[4] = {3, 9, 5, 2} ;
-int butt[5];
+int butt[4];
```
Comment (Antdigapony) [inline LEDGAME.ino:9]: it's better to rename this array to buttState, cause it keeps the state of the buttons and it's a good style to init the array with 0s or 1s

File: LEDGAME.ino [hunk 17]
```diff
@@ -1,12 +1,15 @@
...
+#define GAME_SET_N_START (1)
+#define GAME_RUNNING (2)
+#define GAME_END (3)
```
Comment (Antdigapony) [inline LEDGAME.ino:12]: please, add separate state of the GAME_SCORES where you will show the player's scores for 3-5 seconds

File: LEDGAME.ino [hunk 18]
```diff
@@ -1,12 +1,15 @@
...
 
 void setup()
```
Comment (Antdigapony) [inline LEDGAME.ino:14]: please, move the setup() to the loop(), so in the end, there must be:
- your defines and variables
- your classes and funcs
- setup()
- loop()

File: LEDGAME.ino [hunk 19]
```diff
@@ -20,181 +23,187 @@ void setup()
...
-    if ( (i_timer - old_i_timer) <= limitValue)
-      return true;
-    else 
-      return false;
-  }
+      bool isTimerRunning()
+      {
+        newTimeValue = millis();
+        
+        if ( (newTimeValue - oldTimeValue) <= limitValue)
```
Comment (Antdigapony) [inline LEDGAME.ino:48]: you can simply make the same with:
return ( (newTimeValue - oldTimeValue) <= limitValue )

File: LEDGAME.ino [hunk 20]
```diff
@@ -20,181 +23,187 @@ void setup()
...
+//generating random time for led
+int randomGenTime()
 {
-   digitalWrite(ledInit[num], HIGH);   
+  int randomtime = (rand() % 9 + 3) * 100; 
+  return randomtime;
 }
 
+ //turn on random led
+int randomLedOn(int num)
```
Comment (Antdigapony) [inline LEDGAME.ino:79]: un-correct function's name - this function turns the Led ON - not the random LED, but any LED with number num

File: LEDGAME.ino [hunk 21]
```diff
@@ -20,181 +23,187 @@ void setup()
...
-   digitalWrite(ledInit[2], LOW);
-   digitalWrite(ledInit[3], LOW);
-   digitalWrite(ledInit[0], LOW);
+    digitalWrite(ledInit[num], LOW);
   }  
 
-//выдаем результат в зависимости от нажатой кнопки 
-int buttWait(int num)
+//буфер с нажатыми кнопками
+int WhichButtPressed()
```
Comment (Antdigapony) [inline LEDGAME.ino:92]: please, use one style:
- camelCase
- and start the function names with a lower-case letter

File: LEDGAME.ino [hunk 22]
```diff
@@ -20,181 +23,187 @@ void setup()
...
-        score += 1;
-      }    
-        else if (noButtPress == 1)
-        {
-          Serial.print("                                                          NO BUTTON PRESS DETECTED \n");
-        }
+  
+  for (int c = 1; c < 4; c++)
+  {
+   if (digitalRead(buttInit[c]) == HIGH)
```
Comment (Antdigapony) [inline LEDGAME.ino:97]: you are using both arrays:
- buttonInit[]
- butt

Maybe it is more convenient to change their names to :
- newButtonState[]
- oldButtonState[]

File: LEDGAME.ino [hunk 23]
```diff
@@ -20,181 +23,187 @@ void setup()
...
 }
 
-//очистка буфера
-    void clearbutt()
+//clearing butt bufer
+    void clearButt()
     {
         butt[0] = 1; 
-      for (int i = 1; i < 5; i++)
+      for (int i = 1; i < 4; i++)
```
Comment (Antdigapony) [inline LEDGAME.ino:130]: you can avoid the 'magic number' issues and use a #define NUM_BUTTONS or BUTTONS_QUANTITY

File: LEDGAME.ino [hunk 24]
```diff
@@ -20,181 +23,187 @@ void setup()
...
+   while (butt[0] == 0) 
+   {
+     if(digitalRead(buttInit[0]) == HIGH)
+     { 
+      butt[0] = 1; 
+     }
+   }
 }
+
+void PrintScore()
```
Comment (Antdigapony) [inline LEDGAME.ino:148]: styling - use lower case letter for the functions, please

File: LEDGAME.ino (hunk 21 above)
Comment (Antdigapony) [inline LEDGAME.ino:92]: un-correct name of the function: this function does not return anything and it just sets but[c] = c; if any button is pressed.
Maybe it is better to call it checkAnyButtonPressed() ???
and think about the buttonOldState and buttonNewState

File: LEDGAME.ino [hunk 25]
```diff
@@ -20,181 +23,187 @@ void setup()
...
+     butt[c] = c;
+     i = 1;
+   }
+  } 
+  
+}
+
 
-        return 0;
+int Logic(int num)
```
Comment (Antdigapony) [inline LEDGAME.ino:107]: bad name for the function - Logic does not tell anything what the function does.

maybe call it: checkResult() or checkPressing(). something like that, which will tell what is happening here - in this function

File: LEDGAME.ino [hunk 26]
```diff
@@ -20,181 +23,187 @@ void setup()
...
+{
+ if (butt[num] == num && i == 1)
```
Comment (Antdigapony) [inline LEDGAME.ino:109]: if you are using a global variable as a flag - i
it's better to call what this flag-variable determines

File: LEDGAME.ino [hunk 27]
```diff
@@ -20,181 +23,187 @@ void setup()
...
+  switch (gameState) 
   {
-  case 1: // the game start
+  case GAME_SET_N_START: // the game start
         Serial.print("                                                        Please press Start Button!!!\n\n");
-        buttsWait();
-        timer.setTimer(15000); //сколько будет идти игра
-        timer.updateTimer(); // включаем таймер
-        a = 2;
+        startButtonWait(); //waiting for start button
```
Comment (Antdigapony) [inline LEDGAME.ino:164]: what will happen - if the player will not press the START_BUTTON here ?
will the system react on this ?

Comment (dzhekzzzz) [reply LEDGAME.ino:164]: system will wait for a start_button press for a decades

File: LEDGAME.ino (hunk 19 above)
Comment (dzhekzzzz) [reply LEDGAME.ino:48]: I need to return the value of true \ false, if I change it will this func return it to me?

**Comment by dzhekzzzz** [comment]
changed many bugs
improved work stability
added new cool features
added random time delay between leds blimp
slightly changed design of the information displayed on the screen

for modders: redesigned code syntax (may be easier to read, correct markup and function names)

File: LEDGAME.ino (hunk 19 above)
Comment (Antdigapony) [reply LEDGAME.ino:48]: yes, it will return the result of the condition's checking

File: LEDGAME.ino (hunk 22 above)
Comment (dzhekzzzz) [reply LEDGAME.ino:97]: first arr is only for initing butts, i think its better to call buttInin
second arr is for buffer with butt states, i rename it in buttState

i think these names are better

**Pull request merged**